crates/l2/based/block_fetcher.rs

```diff
@@ -0,0 +1,512 @@
+        loop {
+            let _ = self.main_logic().await.inspect_err(|err| {
+                error!("Block Fetcher Error: {err}");
+            });
+
+            sleep(Duration::from_millis(self.fetch_interval_ms)).await;
+        }
+    }
+
+    pub async fn main_logic(&mut self) -> Result<(), BlockFetcherError> {
+        if let SequencerStatus::Sequencing = self.sequencer_state.status().await {
+            return Ok(());
+        }
+        self.fetch().await
+    }
+
+    async fn fetch(&mut self) -> Result<(), BlockFetcherError> {
+        while !node_is_up_to_date::<BlockFetcherError>(
+            &self.eth_client,
+            self.on_chain_proposer_address,
+            &self.rollup_store,
+        )
+        .await?
+        {
+            info!("Node is not up to date. Syncing via L1");
+
+            let last_l2_block_number_known = self.store.get_latest_block_number().await?;
+
+            let last_l2_batch_number_known = self
+                .rollup_store
+                .get_batch_number_by_block(last_l2_block_number_known)
+                .await?
+                .ok_or(BlockFetcherError::InternalError(format!(
+                    "Failed to get last batch number known for block {last_l2_block_number_known}"
+                )))?;
+
+            let last_l2_committed_batch_number = self
+                .eth_client
+                .get_last_committed_batch(self.on_chain_proposer_address)
+                .await?;
+
+            let l2_batches_behind = last_l2_committed_batch_number.checked_sub(last_l2_batch_number_known).ok_or(
+                BlockFetcherError::InternalError(
+                    "Failed to calculate batches behind. Last batch number known is greater than last committed batch number.".to_string(),
+                ),
+            )?;
+
+            info!("Node is {l2_batches_behind} batches behind. Last batch number known: {last_l2_batch_number_known}, last committed batch number: {last_l2_committed_batch_number}");
+
+            let batch_committed_logs = self.get_logs().await?;
+
+            let mut missing_batches_logs = self
+                .filter_logs(&batch_committed_logs, last_l2_batch_number_known)
+                .await?;
+
+            missing_batches_logs.sort_by_key(|(_log, batch_number)| *batch_number);
+
+            for (batch_committed_log, batch_number) in missing_batches_logs {
+                let batch_commit_tx_calldata = self
+                    .eth_client
+                    .get_transaction_by_hash(batch_committed_log.transaction_hash)
+                    .await?
+                    .ok_or(BlockFetcherError::InternalError(format!(
+                        "Failed to get the receipt for transaction {:x}",
+                        batch_committed_log.transaction_hash
+                    )))?
+                    .data;
+
+                let batch = Self::decode_batch_from_calldata(&batch_commit_tx_calldata)?;
+
+                self.store_batch(&batch).await?;
```

Comment:
This `store_batch` and the call to `seal_batch` below should ideally be a single transactions, both for performance reasons but more importantly to avoid a bug where the node dies in between these two calls and some blocks don't have a batch associated to them because the batch wasn't sealed